File: src/main/java/work/lclpnet/ap2/game/cozy_campfire/setup/CCBaseManager.java
package work.lclpnet.ap2.game.cozy_campfire.setup;

import net.minecraft.entity.Entity;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.math.BlockPos;
import work.lclpnet.ap2.api.game.team.Team;
import work.lclpnet.ap2.api.game.team.TeamManager;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public class CCBaseManager {

    private final Map<Team, CCBase> bases;
    private final TeamManager teamManager;

    public CCBaseManager(Map<Team, CCBase> bases, TeamManager teamManager) {
        this.bases = bases;
        this.teamManager = teamManager;
    }

    private CCBase requireBase(Team team) {
        return Objects.requireNonNull(bases.get(team), "Base not configured for team " + team.getKey().id());
    }

    public boolean isInAnyBase(double x, double y, double z) {
        return bases.values().stream().anyMatch(base -> base.isInside(x, y, z));
    }

    public boolean isInBase(ServerPlayerEntity player) {
        Team team = teamManager.getTeam(player).orElse(null);
        if (team == null) return false;

        CCBase base = requireBase(team);

        return base.isInside(player.getX(), player.getY(), player.getZ());
    }

    public Optional<CCBase> getBase(Team team) {
        return Optional.ofNullable(bases.get(team));
    }

    public Map<Team, CCBase> getBases() {
        return bases;
    }

    public Optional<Team> getCampfireTeam(BlockPos pos) {
        return bases.entrySet().stream()
                .filter(entry -> entry.getValue().getCampfirePos().equals(pos))
                .map(Map.Entry::getKey)
                .findAny();
    }

    public Optional<Team> getEntityTeam(Entity entity) {
        return bases.entrySet().stream()
                .filter(entry -> entry.getValue().isEntity(entity))
                .map(Map.Entry::getKey)
                .findAny();
    }
}

File: src/minigames/cozy_campfire/src/main/kotlin/work/lclpnet/ap2/game/cozy_campfire/setup/CCBaseManager.kt
package work.lclpnet.ap2.game.cozy_campfire.setup

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import work.lclpnet.ap2.game.team.Team
import work.lclpnet.ap2.game.team.TeamManager
import work.lclpnet.kibu.util.StructureWriter
import work.lclpnet.kibu.util.StructureWriter.Option.*
import work.lclpnet.kibu.util.math.Matrix3i
import java.util.*

class CCBaseManager(
    private val bases: Map<Team, CCBase>,
    private val teamManager: TeamManager
) {

    private fun requireBase(team: Team): CCBase =
        checkNotNull(bases[team]) { "Base not configured for team ${team.key.id}" }

    fun isInAnyBase(x: Double, y: Double, z: Double) =
        bases.values.any { it.isInside(x, y, z) }

    fun isInBase(player: ServerPlayer): Boolean {
        val team = teamManager.getTeam(player) ?: return false

        return requireBase(team).isInside(player.x, player.y, player.z)
    }

    fun getBase(team: Team) = bases[team]

    fun getBases(): Map<Team, CCBase> = bases

    fun getCampfireTeam(pos: BlockPos): Team? =
        bases.entries.firstOrNull { it.value.campfirePos == pos }?.key

    fun getEntityTeam(entity: Entity): Team? =
        bases.entries.firstOrNull { it.value.isEntity(entity) }?.key

    fun openDoors(world: ServerLevel) {
        val opts = EnumSet.of(FORCE_STATE, SKIP_DROPS, SKIP_NEIGHBOUR_UPDATE)

        for (base in bases.values) {
            val struct = base.doorSchematic ?: continue
            val pos = base.doorPos ?: continue

            StructureWriter.placeStructure(
                struct,
                world,
                pos,
                Matrix3i.IDENTITY,
                opts
            )
        }
    }
}
